sql/bulksst: coordinate distributed SST metadata merge with CombineFileInfo

spilchen opened the pull request:
Implements CombineFileInfo(), a coordinator utility that aggregates SST metadata from distributed workers and determines merge task spans based on sampled keys.

The function combines SST file metadata from multiple workers and uses their row samples to split schema spans into merge task spans. This will be used by the new distributed merge pipeline.

Resolves: #156662
Epic: CRDB-48845

Release note: none

Co-authored by: @jeffswenson

mw5h reviewed (changes requested):
> @mw5h reviewed 1 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @kev-cao)

___
*[`pkg/sql/bulksst/combine_file_info.go` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157004#-Odu6tPG0z1hPtkNUZ8j:-Odu6tPG0z1hPtkNUZ8k:bbfdxkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/51fed3fc79d305de4980377d7cb8ba95373a6a87/pkg/sql/bulksst/combine_file_info.go#L17)):*
> ```Go
> )
> 
> // CombineFileInfo combines SST file metadata and determines merge task spans based on key samples.
> ```

Perhaps we should describe what we're producing here and why? There's not really an in-code description of what the samples are, so it's not necessarily clear why we want to use them as break points for our merge spans.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/bulksst: coordinate distributed SST metadata merge with CombineFileInfo

Implements CombineFileInfo(), a coordinator utility that aggregates SST
metadata from distributed workers and determines merge task spans based on
sampled keys.

The function combines SST file metadata from multiple workers and uses their
row samples to split schema spans into merge task spans. This will be used by
the new distributed merge pipeline.

Resolves: #156662
Epic: CRDB-48845

Release note: none

Co-authored by: @jeffswenson

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @kev-cao, and @mw5h)

___
*[`pkg/sql/bulksst/combine_file_info.go` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157004#-Odu6tPG0z1hPtkNUZ8j:-Odx9jUR96paW3vFcRyX:b-mpx2e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/51fed3fc79d305de4980377d7cb8ba95373a6a87/pkg/sql/bulksst/combine_file_info.go#L17)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Perhaps we should describe what we're producing here and why? There's not really an in-code description of what the samples are, so it's not necessarily clear why we want to use them as break points for our merge spans.
</blockquote></details>

Good idea. I beefed up the function comment.


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

mw5h reviewed (approved):
LGTM!

> @mw5h reviewed 2 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157004#-:-OeIzuifA1yfkMJ9Fdet:b-85i648)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)



<!-- Sent from Reviewable.io -->
